### PR TITLE
udev: when btrfs.ko is not available consider btrfs filesystems not r…

### PR DESCRIPTION
**Pull request by poettering**
…eady

Let's add a special tweak to the btrfs builtin: if /dev/btrfs-control is
not there, let's consider all btrfs file systems as SYSTEMD_READY=0.
This is useful in initrds, where btrfs.ko might be missing. After the
initrd → host transition we can then retigger the device and undo the
SYSTEMD_READY=0 marking.

### CODE REVIEW

**Review by haraldh — APPROVED** [review]
LGTM

File: src/udev/udev-builtin-btrfs.c [hunk 1]
```diff
@@ -21,8 +21,17 @@ static int builtin_btrfs(sd_device *dev, int argc, char *argv[], bool test) {
                 return log_device_error_errno(dev, SYNTHETIC_ERRNO(EINVAL), "Invalid arguments");
 
         fd = open("/dev/btrfs-control", O_RDWR|O_CLOEXEC);
-        if (fd < 0)
+        if (fd < 0) {
+                if (IN_SET(errno, ENOENT, ENXIO, ENODEV)) {
+                        /* Driver not installed? Then we aren't ready. This is useful in initrds, that lack
+                         * btrfs.ko. After the host transition the device can be retriggerd and will then be
```
Comment (bluca) [inline src/udev/udev-builtin-btrfs.c]: retriggerd/retriggered

**Comment by poettering** [comment]
Force pushed a new version, with the typo fix. No other changes. Upgrading green label.

**Comment by arianvp** [comment]
Hmm

This commit broke booting a btrfs rootfs + systemd-cryptsetup for me and made my system unbootable after updating to 247. I dissected the boot failure to https://github.com/systemd/systemd-stable/commit/3f94149a9bdf7267453f614aa85f1735913d91a2. Reverting the patch fixed booting again.

The commit seems to break the following scenario:

I have the `btrfs.ko` file in `/lib/modules/5.4.104/kernel/fs/btrfs/btrfs.ko.xz`

in initramfs after `/dev/gpt-auto-root-luks` is decrypted it appears as `/dev/dm-0   | /dev/gpt-auto-root`


`sysroot.mount` pulls in `/dev/gpt-auto-root` device unit.   In the previous scenario the lack of `/dev/btrfs-control` wouldn't trigger `SYSTEMD_READY=0`.  so `sysroot.mount` would be reached.  The `mount` command that `sysroot.mount` executes then loads the `btrfs` kernel module; and successfully mounts the filesystem.

Problem with this change is that now `sysroot.mount` never gets reached; and thus the btrfs kernel module is never lazy loaded by the kernel when mount syscall is being called.  Hence the boot gets stuck timing out waiting on `/dev/gpt-auto-root` to become ready.

A workaround is adding `btrfs` to `modules-load.d` but this feels like a hack.

Another workaround is not including this  udev rule in the initrd.

Thoughts?

**Comment by arianvp** [comment]
Maybe `mount` lazy-loading `btrfs` and mounting it immediately is just a broken idea given that we need to probe `btrfs` for readiness first in case of a RAID setup and indeed the only sane way is to preload the btrfs module in `modules-load.d`

@haraldh how does Dracut do this?

**Comment by poettering** [comment]
@arianvp which initrd implementation is this? Is it possibly missing the systemd-tmpfiles-setup-dev.service + kmod-static-nodes.service stuff and such?

normally, /dev/btrfs-control should be created by a node by s-t-s-d.s based on the data k-s-n.s extracts from the .ko metadata. But if one of the two is not included in the initrd this is not going to work.

**Comment by arianvp** [comment]
It is my own. I'm working on bring a systemd-based initrd to NixOS which is currently work in progress. I do have these services included in the initrd but these are good pointers to look if maybe something is going wrong there. Thanks for the pointer.   

https://github.com/arianvp/nixos-stuff/pull/46/files#diff-99dce9c142ff9a95717f2887b335d2eec778600f6569045d5f86aa23aff03060R41

https://github.com/arianvp/nixos-stuff/pull/46/files#diff-99dce9c142ff9a95717f2887b335d2eec778600f6569045d5f86aa23aff03060R128

**Comment by arianvp** [comment]
Thanks @poettering.  the problem was indeed with `kmod-static-nodes.service`  which apparently NixOS patched in a way that made it not execute properly.

**Pull request merged**